### PR TITLE
Added ioc context to inject. Step 6

### PR DESCRIPTION
**Pull request by igorkorsukov**

### CODE REVIEW

**Comment by Jojo-Schmitz** [comment]
Are you going to remove src/framework/global/thirdparty/kors_modularity and just keep  src/framework/global/modularity?
I'm asking because currently they seem to conflict with one another, error messaze from MSVC:
```
Severity	Code	Description	Project	File	Line	Suppression State	Details
Warning	MSB8027	Two or more files with the name of ioc.cpp will produce outputs to the same location. This can lead to an incorrect build result.  The files involved are C:\Users\Jojo\Documents\GitHub\MuseScore\src\framework\global\thirdparty\kors_modularity\modularity\ioc.cpp, C:\Users\Jojo\Documents\GitHub\MuseScore\src\framework\global\modularity\ioc.cpp.	notation	C:\Program Files\Microsoft Visual Studio\2022\Community\MSBuild\Microsoft\VC\v170\Microsoft.CppBuild.targets	1119		
``

**Comment by igorkorsukov** [comment]
> Are you going to remove src/framework/global/thirdparty/kors_modularity and just keep src/framework/global/modularity? I'm asking because currently they seem to conflict with one another, warning message from MSVC: `Warning MSB8027 Two or more files with the name of ioc.cpp will produce outputs to the same location. This can lead to an incorrect build result. The files involved are C:\Users\Jojo\Documents\GitHub\MuseScore\src\framework\global\thirdparty\kors_modularity\modularity\ioc.cpp, C:\Users\Jojo\Documents\GitHub\MuseScore\src\framework\global\modularity\ioc.cpp. `

thank you, fixed

File: src/notation/notationtypes.h [hunk 1]
```diff
@@ -429,6 +429,14 @@ struct StaffConfig
     ClefTypeList clefTypeList;
     engraving::StaffType staffType;
 
+    StaffConfig(const engraving::Color& defaultColor)
```
[the remark this answers is not in the document]
Comment (cbjeukendrup) [reply src/notation/notationtypes.h:432]: Hm, actually, sorry, that probably won't work. We can leave it as it is in this PR, but I do think we should refactor this later (perhaps it would be solved by making `EngravingConfiguration` global). Because I do think it's a bit ugly this way.

**Comment by igorkorsukov** [comment]
> Seeing this PR, I'm still starting to doubt again whether "apply uniform structure everywhere, then look into making things global" is a good strategy. Using non-global injections seems to result in not always super beautiful code (especially that the default color now needs to be passed to the staffType constructor, and all consequences of that; and the changes to `WidgetUtils`). All these "ugly" things will need to remembered and undone later, when we make things global. Wouldn't we make our lives much easier by making things global _now_, so that we don't spend ages on trying to get iocContext in a lot of difficult places where we will soon not need it anymore because interfaces become global?

In general, I agree with you! It's probably not a good idea to blindly follow chosen strategy. 
I will make some injects is global

File: src/engraving/dom/textbase.cpp [hunk 2]
```diff
@@ -741,19 +741,36 @@ int TextCursor::position(int row, int column) const
...
+    text = f.text;
+    format = f.format;
+}
+
+TextFragment& TextFragment::operator =(const TextFragment& f)
+{
+    text = f.text;
+    format = f.format;
+    pos = f.pos;
+    return *this;
```
Comment (cbjeukendrup) [inline src/engraving/dom/textbase.cpp:773]: Why are these necessary?

Comment (igorkorsukov) [reply src/engraving/dom/textbase.cpp:773]: Inject does not have a copy operator, so a compiler cannot create a default implementation of the copy operator

File: src/engraving/dom/stafftype.cpp [hunk 3]
```diff
@@ -686,8 +687,9 @@ String StaffType::tabBassStringPrefix(int strg, bool* hasFret) const
 //    rect    the rect of the 'blue rectangle' showing the input position
 //---------------------------------------------------------
 
-void StaffType::drawInputStringMarks(Painter* p, int string, voice_idx_t voice, const RectF& rect) const
+void StaffType::drawInputStringMarks(Painter* p, int string, voice_idx_t voice, const Color& selectionColor, const RectF& rect) const
 {
+    UNUSED(voice);
```
Comment (cbjeukendrup) [inline src/engraving/dom/stafftype.cpp:692]: Why not just remove `voice`?

File: src/engraving/dom/textbase.h [hunk 4]
```diff
@@ -212,19 +212,25 @@ class TextCursor
 //    contains a styled text
 //---------------------------------------------------------
 
-class TextFragment
+class TextFragment : public muse::Injectable
 {
-    INJECT_STATIC(IEngravingFontsProvider, engravingFonts)
+public:
+    muse::Inject<IEngravingFontsProvider> engravingFonts = { this };
```
Comment (cbjeukendrup) [inline src/engraving/dom/textbase.h:218]: I think this one should also be global

File: src/engraving/dom/stafftype.cpp (hunk 3 above)
Comment (igorkorsukov) [reply src/engraving/dom/stafftype.cpp]: removed

File: src/notation/notationtypes.h [hunk 5]
```diff
@@ -431,6 +431,11 @@ struct StaffConfig
     ClefTypeList clefTypeList;
     engraving::StaffType staffType;
 
+    void setStaffTypeColor(const engraving::Color& val)
```
Comment (cbjeukendrup) [inline src/notation/notationtypes.h:434]: Seems unused

Comment (igorkorsukov) [reply src/notation/notationtypes.h]: removed

File: src/engraving/dom/textbase.h [hunk 6]
```diff
@@ -237,10 +242,12 @@ class TextFragment
...
 //---------------------------------------------------------
 
-class TextBlock
+class TextBlock : public muse::Injectable
 {
 public:
-    TextBlock() {}
+    TextBlock(const muse::modularity::ContextPtr& iocCtx)
+        : muse::Injectable(iocCtx) {}
+
```
Comment (cbjeukendrup) [inline src/engraving/dom/textbase.h:250]: This is now unnecessary

File: src/importexport/musicxml/internal/musicxml/musicxmlfonthandler.cpp [hunk 7]
```diff
@@ -119,7 +119,8 @@ static int plainTextPlusSymbolsListSize(const std::list<TextFragment>& list)
  */
 
 bool MScoreTextToMXML::split(const std::list<TextFragment>& in, const int pos, const int len,
-                             std::list<TextFragment>& left, std::list<TextFragment>& mid, std::list<TextFragment>& right)
+                             std::list<TextFragment>& left, std::list<TextFragment>& mid, std::list<TextFragment>& right,
+                             const muse::modularity::ContextPtr& iocCtx)
```
Comment (cbjeukendrup) [inline src/importexport/musicxml/internal/musicxml/musicxmlfonthandler.cpp:123]: Now iocCtx is unused here (seems to become unnecessary in quite some cases in exportxml.cpp too)

(sorry for the big amount of time between my comments, I'm multitasking a bit)

File: src/engraving/dom/textbase.h (hunk 6 above)
Comment (igorkorsukov) [reply src/engraving/dom/textbase.h]: thank you, removed

File: src/importexport/musicxml/internal/musicxml/musicxmlfonthandler.cpp (hunk 7 above)
Comment (igorkorsukov) [reply src/importexport/musicxml/internal/musicxml/musicxmlfonthandler.cpp]: thank you, removed

**Review by cbjeukendrup — APPROVED** [review]
